discovery-infra/assisted_service_api.py

```diff
@@ -14,9 +18,51 @@ def __init__(self, inventory_url):
         self.inventory_url = inventory_url
         configs = Configuration()
         configs.host = self.inventory_url + "/api/assisted-install/v1"
+        if 'OFFLINE_TOKEN' in os.environ:
+            self.set_config_auth(configs)
+
         self.api = ApiClient(configuration=configs)
         self.client = api.InstallerApi(api_client=self.api)
 
+    def set_config_auth(self, c):
+        offline_token = os.environ.get('OFFLINE_TOKEN')
+        if offline_token == None: return
+
+        def refresh_api_key(config):
+            # Get the properly padded key segment
+            auth = config.api_key.get('Authorization', None)
+            if auth != None:
+                segment = auth.split('.')[1]
+                padding = len(segment) % 4
+                segment = segment + padding * '='
+
+                expires_on = json.loads(base64.b64decode(segment))['exp']
+
+                # if this key doesn't expire or if it has more than 30 seconds left, don't refresh
+                remaining = expires_on - time.time()
+                if expires_on == 0 or remaining > 30:
+                    return
+
+            # fetch new key if expired or not set yet
+            token_url = "https://sso.redhat.com/auth/realms/redhat-external/protocol/openid-connect/token"
+            params = {
+                "client_id":     "cloud-services",
+                "client_secret": offline_token,
+                "grant_type":    "refresh_token",
+                "refresh_token": offline_token,
+            }
+
+            log.info("Refreshing API key")
+            response = requests.post(token_url, data = params)
```

Comment:
should we also catch exceptions?

Reply:
Sorry, I'm not super familiar with python yet. What exceptions could this raise? None of the docs/examples I was reading mentioned that.

Reply:
i think u can catch all exception and do log.exception with nice message

Reply:
on the other hand, what to do if exception occurs? exit the test?

Reply:
Yeah, if we fail here then the test will just fail. I mean we could continue and let auth fail, but I think that would be harder to debug.

We know here that the test will not succeed so better to throw an exception or log that we failed to fetch the token and fail here.

Reply:
agree, so lets throw